### PR TITLE
feat(frontend): support InSubquery.

### LINKED ISSUE #2027: frontend: support IN subquery
**Is your feature request related to a problem? Please describe.**
TPC-H Q20 uses IN subquery, so we need to support it.

### PR DESCRIPTION
**Pull request by likg227**
## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
In this pr, I supported (NOT) InSubquery for WHERE clause and add `on` for `LogicalApply`.
I bind InSubquery as normal InList and plan it as Left-Semi-Apply using equal condition as its predicate.

- Describe any limitations of the current code (optional)
Please note that InSubquery that goes into others cannot be handled by substitute_subqueries yet.
## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
close #2027

### CODE REVIEW

File: src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@
   logical_plan: |
     LogicalProject { exprs: [$1, $2], expr_alias: [x, y] }
       LogicalFilter { predicate: ($1 > $3) }
-        LogicalApply { type: LeftOuter }
+        LogicalApply { type: LeftOuter, on: always }
```
Comment (fuyufjh) [inline src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:8]: Recommend using `TRUE` instead of `always`. May fix it later.

File: src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml [hunk 2]
```diff
@@ -151,3 +151,20 @@
                 LogicalScan { table: t1, columns: [_row_id#0, x, y] }
                 LogicalProject { exprs: [1:Int32, $0], expr_alias: [1,  ] }
                   LogicalScan { table: t2, columns: [y] }
+- sql: |
+    create table t1(x int, y int);
+    create table t2(x int, y int);
+    select x from t1 where y in (select y from t2 where t1.x = t2.x);
```
Comment (fuyufjh) [inline src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: Can you add more tests? e.g.

- `not in`
- with other conditions in the subquery
- with more than 1 correlated variable in the subquery
- with non-equi correlated condition in the subquery
- ...

File: src/frontend/src/planner/select.rs [hunk 3]
```diff
@@ -101,56 +101,39 @@ impl Planner {
...
-                    }
+                    ExprImpl::FunctionCall(func_call) => match func_call.get_expr_type() {
+                        ExprType::In if matches!(func_call.inputs()[1], ExprImpl::Subquery(_)) => 0,
+                        ExprType::Not => {
+                            let expr = &func_call.inputs()[0];
+                            match expr {
+                                ExprImpl::Subquery(_) => 1,
+                                ExprImpl::FunctionCall(func)
+                                    if func.get_expr_type() == ExprType::In
+                                        && matches!(func.inputs()[1], ExprImpl::Subquery(_)) =>
```
Comment (xiangjinwu) [inline src/frontend/src/planner/select.rs:112]: Maybe we can simply the matching by updating the definition of `expr::Subquery`:

Option A:
```
pub struct Subquery {
    pub query: BoundQuery,
    pub kind: SubqueryKind,
    pub negated: bool, // only meaningful for Existential or SetComparison, maybe Option<bool>
    pub operand: Option<ExprImpl>,  // only meaningful for SetComparison
}
```

Option B:
```
pub enum SubqueryKind {
    Scalar,
    Existential(bool),
    SetComparison(bool, ExprImpl),
}
```

This way we avoid packing and unpacking with `ExprType::In` or `ExprType::Not`.

Note that this is still not enough to represent `a >= ALL(...)` or `a <> SOME(...)` yet. The comparison operator also needs to be stored, as well as differentiating `ALL` from `SOME`.

File: src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml (hunk 2 above)
Comment (likg227) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: I have added tests for `not in` in `subquery_expr.yaml`, because we cann't unnest LogicalAntiApply now. I will add other types of tests.

File: src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml (hunk 1 above)
Comment (likg227) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:8]: This behavior is aligned with `LogicalJoin`. Maybe we can discuss it later.

File: src/frontend/src/planner/select.rs [hunk 4]
```diff
@@ -166,6 +149,65 @@ impl Planner {
...
+                let subquery = inputs.pop().unwrap().into_subquery().unwrap();
+                let is_correlated = subquery.is_correlated();
+
+                let left = inputs.pop().unwrap();
+                let right = InputRef::new(input.schema().fields().len(), subquery.return_type());
+                let eq_cond = FunctionCall::new(ExprType::Equal, vec![left, right.into()])?;
+
+                let right_child = self.plan_query(subquery.query)?.as_subplan();
+
+                *input = Self::create_apply_or_join(
```
Comment (xiangjinwu) [inline src/frontend/src/planner/select.rs:181]: Variable names are a little bit hard to follow here.
* `inputs`, `left`, `right` are about exprs
* `input`, `right_child` are about plan nodes.

Maybe we can call them `left_expr` (or `operand`), `right_expr`, `right_plan`.

File: src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml (hunk 2 above)
Comment (fuyufjh) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: It's okay. You may add a test with error output if that's not supported yet.

File: src/frontend/src/planner/select.rs [hunk 5]
```diff
@@ -101,56 +101,39 @@ impl Planner {
...
+                                ExprImpl::FunctionCall(func)
+                                    if func.get_expr_type() == ExprType::In
+                                        && matches!(func.inputs()[1], ExprImpl::Subquery(_)) =>
+                                {
+                                    1
+                                }
+                                _ => 2,
+                            }
+                        }
+                        _ => 2,
```
Comment (xiangjinwu) [inline src/frontend/src/planner/select.rs:119]: In the description of this PR, it is better to be clear about the limitation:
in-subquery that goes into `others` cannot be handled by `substitute_subqueries` yet.

File: src/frontend/src/planner/select.rs (hunk 4 above)
Comment (xxchan) [reply src/frontend/src/planner/select.rs:181]: or use a block
```rust
let (_, inputs, _) = func_call.decompose();
    let (e, subquery) = inputs.into_iter().collect_tuple().unwrap();
    let subquery = subquery.into_subquery().unwrap();
    let is_correlated = subquery.is_correlated();

    let eq_cond = {
        let left = e;
        let right =
            InputRef::new(input.schema().fields().len(), subquery.return_type());
        FunctionCall::new(ExprType::Equal, vec![left, right.into()])?
    };
```

File: src/frontend/src/planner/select.rs (hunk 3 above)
Comment (xxchan) [reply src/frontend/src/planner/select.rs:112]: +1

Comment (xxchan) [reply src/frontend/src/planner/select.rs:112]: An additional reason is that calling `q` in `e in q` "SetComparison" looks a little strange.

File: src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml (hunk 2 above)
Comment (xxchan) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: +1 for an error test

Comment (likg227) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: I just tried it and it will panic, so it doesn't display any error. I just asked @skyzh about this, and he said that planner test can't handle panic.

Comment (skyzh) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]:
```rust
impl ToStream for LogicalApply {
    fn to_stream(&self) -> PlanRef {
        panic!("LogicalApply should be unnested")
    }

    fn logical_rewrite_for_stream(&self) -> (PlanRef, ColIndexMapping) {
        panic!("LogicalApply should be unnested")
    }
}
```

Will panic instead of returning an error. Planner test won't handle panic for now. If we want to test unsupported cases, we will need to refactor this into an error.

Comment (fuyufjh) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: I would recommend to returns an `Err` to user instead of `panic!` or `todo!`. 

As far as I know, no products can unnest all kinds of corrected subqueries, so we **have** to tell users "this query is not supported balabala"

Comment (st1page) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: +1, I think we can do this in future PR.

File: src/frontend/src/planner/select.rs (hunk 3 above)
Comment (likg227) [reply src/frontend/src/planner/select.rs:112]: IMO, refining the definition of `Subquery` or `SubqueryKind` deserves a new pr. As for these two options, extending `SubqueryKind` looks better for me, but it still causes problems when do the matching because we need to extract `SubqueryKind` from `Subquery` now. In conclusion, defining a good struct for Subquery is not that simple.

File: src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml (hunk 2 above)
Comment (likg227) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: We have found that create_apply_or_join is ok to add a check. I will add an error report there.😇😇😇😇

Comment (st1page) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: rethinking the issue, I consider we can do handle the unsupported sql in binder and planer and in optimizer we can panic. for the unnest case, I think `create_apply_or_join` is a good place in planner to return error

Comment (xiangjinwu) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: `create_apply_or_join` may not be enough. This issue is mostly specific to `ExprImpl::Subquery`, `ExprImpl::CorrelatedInputRef`, and `LogicalApply`: they are supposed to be present in the expr tree or plan tree in early phases, and completely replaced later. Another way to report errors is to add a dedicate phase than scans the whole plan for any undesired residuals - if we do not want to make existing `to_prost` / `to_batch` to return `Result`.

Comment (likg227) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: >  Another way to report errors is to add a dedicate phase than scans the whole plan for any undesired residuals

Do you mean do this check right after unnesting? If so, then we should add optimizer_error.

Comment (xiangjinwu) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: I meant impl `PlanVisitor` just to check the existence of one of the three, and call it in `PlanRoot::gen_batch_query_plan` / `PlanRoot::gen_stream_plan` immediately after `gen_optimized_logical_plan`.

Comment (likg227) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: In conclusion, we will add optimizer error or something else in future prs to report error instead of panicing.

**Review by xiangjinwu — APPROVED** [review]
LGTM!

Just a small note for future considerations:
* `NOT (a NOT IN subquery)` may cause a problem ...
* It may be simpler to use same logic of `negated` as `NOT EXISTS`.

**Comment by likg227** [comment]
> LGTM!
> 
> Just a small note for future considerations:
> 
> * `NOT (a NOT IN subquery)` may cause a problem ...
> * It may be simpler to use same logic of `negated` as `NOT EXISTS`.

NOT EXISTS will be parsed to NOT + EXISTS. They are not an entity.

File: src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml (hunk 2 above)
Comment (st1page) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:157]: > impl PlanVisitor just to check the existence of one of the three
+1
